### PR TITLE
create setter for inventory_space

### LINKED ISSUE #142: inventory_space() misses a setter
**Problem summary/missing feature:**  
board_items.Immovable implements the `inventory_space()` method but there is no setter and users needs to use the private variable (undocumented).

**Expected behavior:**  
We need an easy and documented way to set inventory space after construction.

**Work to do:**  
transform `inventory_space()` into a `@property` that manages the private variable `_inventory_space`.

### PR DESCRIPTION
**Pull request by morrme**
## Description

Converted inventory_space() to a @property and added a setter.

Fixes #142 

## Type of change

Please delete options that are not relevant.
(not sure which of these would apply)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by morrme** [comment]
@arnauddupuis 

Codecov is complaining about coverage for this line: 
https://github.com/morrme/pygamelib/blob/issue-142/pygamelib/engine.py#L2745

I created a test using that method and condition here:

https://github.com/morrme/pygamelib/blob/issue-142/tests/test_game.py#L295

..but coverage has not changed. Any advice? 

Thanks!

**Comment by arnauddupuis** [comment]
> @arnauddupuis
> 
> Codecov is complaining about coverage for this line:
> https://github.com/morrme/pygamelib/blob/issue-142/pygamelib/engine.py#L2745
> 
> I created a test using that method and condition here:
> 
> https://github.com/morrme/pygamelib/blob/issue-142/tests/test_game.py#L295
> 
> ..but coverage has not changed. Any advice?
> 
> Thanks!

Hum, this is weird... I'll pull your code and have a look. I had some cases like that where the coverage was not correctly detected. 
The dirty solution is to exclude the line from coverage tests, but I've been removing these instances a lot recently (fixing tests instead).
I'll have a look tonight.

**Comment by arnauddupuis** [comment]
I think that you are incredibly unlucky on that one...
It looks like your code "benefits" from a side effect of an error that has nothing to do with you...

Change the code of _obj2ref() like that:
```python
        elif isinstance(obj, board_items.Door):
            ref["value"] = obj.value
            ref["inventory_space"] = obj.inventory_space()
            ref["overlappable"] = obj.overlappable()
            ref["pickable"] = obj.pickable()
            ref["restorable"] = obj.restorable()
        elif isinstance(obj, board_items.GenericActionableStructure) or isinstance(
            obj, board_items.GenericStructure
        ):
            ref["value"] = obj.value
            ref["inventory_space"] = obj.inventory_space()
            ref["overlappable"] = obj.overlappable()
            ref["pickable"] = obj.pickable()
            ref["restorable"] = obj.restorable()
```
And your coverage should be restored. I looked at your code quickly and it looks fine to me, but I suspect the volume of code added tilted the balance and the reason coverage is complaining is because it's below the coverage threshold for a diff.
The Door code was never tested because it was never reached: a Door is a GenericStructure therefore all Door objects are converted as GenericStructure (and the obvious conclusion is that the code for Door is useless). The issue is simply a question of elif order.

This should provide the required coverage for your code to pass the test.
I haven't reviewed your code yet, I'll do that next.

File: tests/test_game.py [hunk 1]
```diff
@@ -286,6 +286,13 @@ def test_tools_function(self):
         self.assertEqual(g._string_to_constant("DRDOWN"), constants.DRDOWN)
         self.assertEqual(g._string_to_constant("DLUP"), constants.DLUP)
         self.assertEqual(g._string_to_constant("DLDOWN"), constants.DLDOWN)
+        test_wall = g._obj2ref(board_items.Wall(inventory_space=1))
+        self.assertEqual(test_wall["name"], "wall")
+        self.assertEqual(test_wall["inventory_space"], 1)
+        test_door = g._obj2ref(board_items.Door(value=10, inventory_space=1))
+        self.assertEqual(test_door["name"], "Door")
+        self.assertEqual(test_door["value"], 10)
+        self.assertEqual(test_door["inventory_space"], 1)
```
Comment (arnauddupuis) [inline tests/test_game.py:295]: Testing specifically _obje2ref() should be useless now (after you change the order of the elif in _obj2ref) but it probably doesn't hurt.

File: pygamelib/board_items.py [hunk 2]
```diff
@@ -1066,6 +1066,7 @@ def can_move(self):
         """
         return False
 
+    @property
     def inventory_space(self):
         """Return the size of the Immovable Item for the
```
Comment (arnauddupuis) [inline pygamelib/board_items.py:1071]: What do you think of:
```
Return the size that the Immovable item takes in the :class:`~pygamelib.engine.Inventory`.
```
?

**Review by arnauddupuis — CHANGES_REQUESTED** [review]
I know it's not your bug but If you could change the order of the elifs in _obj2ref() it would be very nice ;)

File: pygamelib/board_items.py (hunk 2 above)
Comment (morrme) [reply pygamelib/board_items.py:1071]: > What do you think of:
> 
> ```
> Return the size that the Immovable item takes in the :class:`~pygamelib.engine.Inventory`.
> ```
> 
> ?

That works! I will work on all of these changes in a few hours, if that is okay.

**Review by arnauddupuis — APPROVED** [review]

**Comment by arnauddupuis** [comment]
nice... checks are stuck once again. Your PR looks like it is good to merge. Can you confirm.
Maybe one last thing before merging would be to update the docs/source/authors.rst file and add yourself to it.

**Comment by morrme** [comment]
> nice... checks are stuck once again. Your PR looks like it is good to merge. Can you confirm.
> Maybe one last thing before merging would be to update the docs/source/authors.rst file and add yourself to it.

Confirmed and Done!

**Pull request merged**